### PR TITLE
CPLAT-6121 Add OverReact Redux Docs

### PR DESCRIPTION
**Pull request by joebingham-wk**
## Motivation
  
To accompany the rollout of OverReact Redux, documentation is needed to illustrate how to use the new APIs and components
## Changes
  
- Add a new markdown file
- Move the example to the web directory
  - Refactor the example a little to illustrate different options and be practices.
## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 
@kealjones-wk @aaronlademann-wf @greglittlefield-wf @sydneyjodon-wk 
### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
      - Ensure the examples load and the devtools work
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: OverReact_Redux_Documentation.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+- `pure` if `true` (default), connect performs several equality checks that are used to avoid unnecessary
+calls to `mapStateToProps`, `mapDispatchToProps`, `mergeProps`, and ultimately to `render`. These include
+`areStatesEqual`, `areOwnPropsEqual`, `areStatePropsEqual`, and `areMergedPropsEqual`.
+While the defaults are probably appropriate 99% of the time, you may wish to override them with custom
+implementations for performance or other reasons.
+- `forwardRef` if `true`, the `ref` prop provided to the connected component will be return the wrapped component.
+
+  For more info see: https://react-redux.js.org/api/connect#connect
+  
+## Using Multiple Stores
```
Comment (aaronlademann-wf) [inline OverReact_Redux_Documentation.md]: Should we just link to here from the `context` section above - and move the examples, etc into this section?

File: OverReact_Redux_Documentation.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+### Advantages
+
+A full list of advantages can be found [here](https://react-redux.js.org/introduction/why-use-react-redux). 
+Especially useful for OverReact though, OverReact Redux allows for targeted updates that only update components 
+that receive data which has changed, rather than the whole component tree (as is the behavior with React). By utilizing 
+the `connect()` function in conjunction with `mapStateToProps()`, a component will only update when a piece of information 
+it uses is updated.
+
+
+## Using it in your project
```
Comment (aaronlademann-wf) [inline OverReact_Redux_Documentation.md]: This document should have a big bold callout near the beginning that explains how to serve / play around with our functioning demo component.

File: pubspec.yaml [hunk 3]
```diff
@@ -38,10 +38,16 @@ dev_dependencies:
...
 
 dependency_overrides:
   react:
     git:
       url: https://github.com/cleandart/react-dart.git
       ref: 5.1.0-wip
+  socketcluster_client:
+    git:
+      url: https://github.com/kealjones-wk/socketcluster_client.git
+      ref: auto-platform-selection
```
Comment (aaronlademann-wf) [inline pubspec.yaml]: If we can't get this merged into `socketcluster_client`, we may need to move the redux demo into the example dir, and make the example have its own pubspec, etc - so we can place the dependency override there.

File: OverReact_Redux_Documentation.md (hunk 1 above)
Comment (joebingham-wk) [reply OverReact_Redux_Documentation.md]: I like that idea, it's really tightly coupled and the example fits better in the multiple store section.

File: pubspec.yaml (hunk 3 above)
Comment (joebingham-wk) [reply pubspec.yaml]: 👍 That makes sense, I wasn't sure the best way to handle having the PR in progress. Since `3.1.0-wip` isn't released yet, would we keep this here until we look at releasing (or it gets merged) and evaluate then?

Comment (joebingham-wk) [reply pubspec.yaml]: 🎉 🎉 🎉 It merged!! 🎉 🎉 🎉

File: doc/over_react_redux_documentation.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+## Demo Component
+A basic demo component is available in the `web` directory. In addition to running the app to see how it works, the 
+[store](../web/over_react_redux/store.dart) and [component](../web/over_react_redux/components/counter.dart) have 
+comments that provide additional explanation.
+
+To run and experiment with the demo:
+1. `pub get`
+1. `pub run build_runner serve web`
+1. Navigate to `localhost:8080/over_react_redux`
```
Comment (kealjones-wk) [inline doc/over_react_redux_documentation.md]:
```suggestion
1. Navigate to `localhost:8080/over_react_redux/`
```
because your directions don't use webdev_proxy it needs the ending `/` to open :/

Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Good catch!

File: doc/over_react_redux_documentation.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        * [context](#context)
+        * [pure](#pure)
+        * [forwardRef](#forwardref)
+* __[Using Multiple Stores](#using-multiple-stores)__
+* __[Using Redux DevTools](#using-redux-devtools)__
+
+## Purpose
+[React Redux](https://react-redux.js.org) is a UI binding library that allows for better Redux integration with React. 
+This integration allows for APIs that align more closely with React design patterns while also providing more direct 
+access to store data. OverReact Redux is the Dart wrapper around that library.
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Yes sir, I like it!

File: doc/over_react_redux_documentation.md [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    - Called with dispatch as the first argument.
+    
+        You can make use of this by returning new functions that call dispatch() inside themselves,
+and either pass in a plain action directly or pass in the result of an action creator.
+If you need access to the props provided to the connected component you can use `mapDispatchToPropsWithOwnProps`,
+the second argument will be `ownProps`.
+
+        > See: <https://react-redux.js.org/using-react-redux/connect-mapdispatch#defining-mapdispatchtoprops-as-a
+        -function>
+- #### `mergeProps`
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Good catch 😆

File: doc/over_react_redux_documentation.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+           (ReduxProvider()
+             ..store = store2
+             ..context = bigCounterContext
+           )(
+             // ... connected componentry
+           ),
+         )
+    ```
+
+## Using Redux DevTools
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: I definitely agree that the instructions for this is basically is the repo docs, and I should have linked directly to that. So that has been added!

I also see the step by step being verbose when that single copy and paste could work, but I feel like the step by step gives valuable clarity to each independent piece of the update and grouping the segment talking about adding the actual devtools code wouldn't made a huge difference, so I'd rather focus on each part of the addition. What do you think?

[the remark this answers is not in the document]
Comment (kealjones-wk) [reply doc/over_react_redux_documentation.md]: agreement!

File: doc/over_react_redux_documentation.md [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+For this purpose, OverReact Redux has the `ReduxProvider` component. It is required to take in a store instance, and can
+ optionally take in a context object when using multiple stores.
+ 
+ __Example:__
+ ```dart
+react_dom.render(
+  (ReduxProvider()..store = fooStore)(
+    // ... React component tree with connected components
+  )
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Good catch!

File: doc/over_react_redux_documentation.md [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+```
+    
+## `connect`
+### `Overview`
+A wrapper around the JS react-redux `connect` function that supports OverReact components.
+
+__Example:__
+```
+UiFactory<CounterProps> ConnectedCounter = connect<CounterState, CounterProps>(
+        mapStateToProps: (state) => (
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: 🤦‍♂

File: doc/over_react_redux_documentation.md [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    var remoteDevtools = RemoteDevToolsMiddleware('127.0.0.1:8000');
+    
+    Future initDevtools() async {
+      remoteDevtools.store = store;
+      return remoteDevtools.connect();
+    }
+    ```
+1. Near the top of your main.dart file call your `initDevtools`.
+    ```dart
+     initDevtools();
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Really good catch!

[the remark this answers is not in the document]
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: Not necessarily. We found that await could cause confusing behavior because if the consumer does not open the devtools, the the page will never load. Emitting the `await` there allows them to still use to page without having to open up the tools.

File: web/over_react_redux/examples/simple/index.dart [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import './store.dart';
+
+main() {
+  setClientConfiguration();
+
+  react_dom.render(
+    ErrorBoundary()(
+      (ReduxProvider()..store = store)(
+        Dom.div()(
+          Dom.h2()('ConnectedBigCounter Store2'),
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply web/over_react_redux/examples/simple/index.dart]: Yep, I missed cleaning that up!

File: web/over_react_redux/examples/simple/index.dart [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+          ConnectedBigCounter()(
+            Dom.h2()('ConnectedCounter Store1'),
+            ConnectedCounter()(),
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply web/over_react_redux/examples/simple/index.dart]: I agree it's confusing. It can be nested inside or it can not be, but for the purpose of the examples I believe you are correct that it shouldn't be.

File: README.md [hunk 13]
```diff
@@ -8,6 +8,8 @@
 [![Test Coverage](https://codecov.io/github/Workiva/over_react/coverage.svg?branch=master)](https://codecov.io/github/Workiva/over_react?branch=master)
 
 > A library for building statically-typed React UI components using Dart.
+>
+> This library also exposes OverReact_Redux, which has [its own documentation](doc/over_react_redux_documentation.md).
```
Comment (greglittlefield-wf) [inline README.md]: There's no underscore in the name. Italics might also be nice to help separate it from the rest of the text

```suggestion
> This library also exposes _OverReact Redux_, which has [its own documentation](doc/over_react_redux_documentation.md).
```

File: doc/over_react_redux_documentation.md [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+        - You _do not_ use the `mapDispatchToProps` parameter on `connect`.
+        - You _do_ use `mapDispatchToProps` but want access to the `props.dipsatch` function. When using 
+        `mapDispatchToProps`, by default the `dispatch` property on `props` is removed. It can be added back 
+        using `ConnectPropsToMixin`. 
+    ```dart
+    // AppState is a class that represents the application's state and can be defined in the same file as the store.
+    UiFactory<FooProps> ConnectedFoo = connect<AppState, FooProps>()(Foo);
+    
+    @Factory()
+    UiFactory<FoodProps> Foo = _$Foo;
```
Comment (greglittlefield-wf) [inline doc/over_react_redux_documentation.md]: #nit typo
```suggestion
    UiFactory<FooProps> Foo = _$Foo;
```

File: doc/over_react_redux_documentation.md [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+## `connect`
+### Overview
+A wrapper around the JS react-redux `connect` function that supports OverReact components.
+
+> See: <https://react-redux.js.org/api/connect>
+
+__Example:__
+```dart
+UiFactory<CounterProps> ConnectedCounter = connect<CounterState, CounterProps>(
+    mapStateToProps: (state) => (
```
Comment (greglittlefield-wf) [inline doc/over_react_redux_documentation.md]: #nit these should have two space indents since they use trailing commas

File: doc/over_react_redux_documentation.md [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+      Counter()..count = state.count
+    ),
+    mapDispatchToProps: (dispatch) => (
+      Counter()..increment = () => dispatch(IncrementAction())
```
Comment (greglittlefield-wf) [inline doc/over_react_redux_documentation.md]: #nit best practice is to wrap arrow functions in parens so they don't mess up the cascade
```suggestion
      Counter()..increment = (() => dispatch(IncrementAction()))
```

File: doc/over_react_redux_documentation.md [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+
+> See: <https://stackoverflow.com/questions/33619775/redux-multiple-stores-why-not>
+
+__Multiple Stores Example:__
+```dart
+Store store1 = new Store<CounterState>(counterStateReducer, initialState: new CounterState(count: 0));
+Store store2 = new Store<BigCounterState>(bigCounterStateReducer, initialState: new BigCounterState(bigCount: 100));
+
+UiFactory<CounterProps> ConnectedCounter = connect<CounterState, CounterProps>(
+      mapStateToProps: (state) => (Counter()..count = state.count)
```
Comment (greglittlefield-wf) [inline doc/over_react_redux_documentation.md]: #nit would be nice to have this be a two-space indent with a trailing comma to match the one below
```suggestion
    mapStateToProps: (state) => (Counter()..count = state.count),
```

File: doc/over_react_redux_documentation.md [hunk 18]
```diff
@@ -0,0 +1,320 @@
+# OverReact Redux
+> A Dart React Redux wrapper for targeted state updates.
```
Comment (greglittlefield-wf) [inline doc/over_react_redux_documentation.md]: #nit reword this to reduce ambiguity around the grouping of the words "Dart React Redux"
```suggestion
> A Dart wrapper for React Redux, providing targeted state updates.
```

File: doc/over_react_redux_documentation.md (hunk 14 above)
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: 🥘

File: doc/over_react_redux_documentation.md [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+Examples are available in the `web` directory. Each example illustrates a different variation or use case of OverReact
+ Redux. Additionally, the store files contain comments that call out specifics pertaining to that example and 
+ provides further explanation.
+ 
+### Running the Examples
+To run and experiment with the demo:
+1. `pub get`
+1. `pub run build_runner serve web`
+1. Navigate to `localhost:8080/over_react_redux/`
+
```
Comment (kealjones-wk) [inline doc/over_react_redux_documentation.md]: Should we mention a way to see the isolated updates based on the "mapStateToProps" you check/turn on "Highlight updates when components render." from the react dev tools settings.
Screen Shot 2019-10-31 at 10 57 53 AM

File: web/over_react_redux/examples/simple/store.dart [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  Action({this.type, this.value});
+
+  final String type;
+  final dynamic value;
+
+  /// Used to encode the data structure for the Redux DevTools.
+  ///
+  /// The DevTools expect plain JavaScript objects with the `type` property. To
+  /// facilitate that, a method such as [toJson] is necessary.
+  /// > See: <https://github.com/MichaelMarner/dart-redux-remote-devtools#encoding-actions-and-state>
```
[the remark this answers is not in the document]
Comment (joebingham-wk) [reply web/over_react_redux/examples/simple/store.dart]: Good catch!

File: doc/over_react_redux_documentation.md (hunk 19 above)
Comment (joebingham-wk) [reply doc/over_react_redux_documentation.md]: 💯 Good idea! I'll add that

**Comment by joebingham-wk** [comment]
@Workiva/release-management-pp

**Comment by dancleereman-wf** [comment]
@joebingham-wk - can you get a posthumous code review on the last commit? (741bf84)

Rosie is having an issue with the timestamp between the last commit and when @kealjones-wk approved the changes.  This usually happens when the review is opened before the commit is made.

**Comment by kealjones-wk** [comment]
+10
